### PR TITLE
WIP: Use page objects for user-sidebar-test

### LINKED ISSUE #776: Rewrite integration tests for user-sidebar component to use page objects
Falls under umbrella issue #697.

## Resources

- What's a page object? [[1](http://martinfowler.com/bliki/PageObject.html),[2](https://wyeworks.com/blog/2015/5/13/using-the-page-object-pattern-with-ember-cli/)]
- [Quickstart guide](http://ember-cli-page-object.js.org/docs/v1.6.x/quickstart)
- See [`tests/pages/components/donations/create-donation.js`](https://github.com/code-corps/code-corps-ember/blob/develop/tests/pages/components/donations/create-donation.js) and [`tests/integration/components/donations/create-donation-test.js`](https://github.com/code-corps/code-corps-ember/blob/develop/tests/pages/components/donations/create-donation.js) for in-house examples of, respectively, a page object and the page object being utilized in an integration test.
- To generate a page object, run `ember generate page-object component/<component-name>` in the project directory.

### PR DESCRIPTION
**Pull request by sarupbanskota**
Closes #776 
Also under #697

### CODE REVIEW

File: tests/integration/components/user-sidebar-test.js [hunk 1]
```diff
@@ -1,22 +1,29 @@
...
   }
 });
 
 test('it renders', function(assert) {
   assert.expect(1);
-  this.render(hbs`{{user-sidebar}}`);
+  page.render(hbs`{{user-sidebar}}`);
 
-  assert.equal(this.$('.user-sidebar').length, 1, 'Component\'s element is rendered');
+  assert.equal(page.userSidebarCount, 1, 'Component\'s element is rendered');
```
Comment (sarupbanskota) [inline tests/integration/components/user-sidebar-test.js]: Not quite sure why this change isn't behaving how I expected it to. Ideas?

File: tests/pages/components/user-sidebar.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { count, text, isHidden } from 'ember-cli-page-object';
+
+export default {
+  scope : '.user-sidebar',
```
Comment (sarupbanskota) [inline tests/pages/components/user-sidebar.js]: could it be because `count('.user-sidebar')` will look for `.user-sidebar` _within_ the scope (which is `.user-sidebar`)?

File: tests/integration/components/user-sidebar-test.js (hunk 1 above)
Comment (begedin) [inline tests/integration/components/user-sidebar-test.js]: We ought to be removing the "it renders" tests altogether, so feel free to remove this one.

File: tests/integration/components/user-sidebar-test.js [hunk 3]
```diff
@@ -44,29 +55,29 @@ test('it renders all required elements', function(assert) {
...
+  page.render(hbs`{{user-sidebar user=user}}`);
 
-  assert.equal(this.$('.user-sidebar').length, 1, 'Component\'s element is rendered');
-  assert.equal(this.$('h2 .name').text(), 'Josh Smith', 'Their name renders');
-  assert.equal(this.$('h2 .username').text(), 'JoshSmith', 'Their username renders');
-  assert.equal(this.$('li.twitter').text().trim(), '@joshsmith', 'Their twitter handler renders');
-  assert.equal(this.$('li.twitter a').attr('href'), 'https://twitter.com/joshsmith', 'The twitter link renders');
-  assert.equal(this.$('li.website').text().trim(), 'https://codecorps.org', 'Their website renders');
-  assert.equal(this.$('li.website a').attr('href'), 'https://codecorps.org', 'The website link renders');
+  assert.equal(page.userSidebarCount, 1, 'Component\'s element is rendered');
```
Comment (begedin) [inline tests/integration/components/user-sidebar-test.js]: We don't really need this assertion at all.

File: tests/pages/components/user-sidebar.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { isVisible, text, isHidden, property } from 'ember-cli-page-object';
+
+export default {
+  scope : '.user-sidebar',
+
+  userSidebarCount    : isVisible(),
```
Comment (begedin) [inline tests/pages/components/user-sidebar.js]: I don't think we need this property. I would name it `userSidebarVisible` or `userSidebarRendered` in context, but if the removal of the "it renders" test above and the one assertion in the other test means we no longer use this property, just remove it from here.

File: tests/pages/components/user-sidebar.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  name                : text('h2 .name'),
+  username            : text('h2 .username'),
+  twitterHandle       : text('li.twitter'),
+  website             : text('li.website'),
+  twitterHandleHidden : isHidden('li.twitter'),
+  websiteHidden       : isHidden('li.website'),
+  twitterLink         : property('href', 'li.twitter a'),
+  websiteLink         : property('href', 'li.website a')
```
Comment (begedin) [inline tests/pages/components/user-sidebar.js]: There are arguments for this type of formatting/spacing, but we don't use it anywhere in the app ourselves, so we would prefer if you switched this to just using a single space instead of horizontally aligning the values.

**Comment by sarupbanskota** [comment]
@begedin thanks for the feedback! NY time so a bit slow, but I should clear this up in a day or two.

File: tests/pages/components/user-sidebar.js (hunk 5 above)
Comment (sarupbanskota) [inline tests/pages/components/user-sidebar.js]: alright cool. I was actually supposed to introduce that with https://github.com/code-corps/code-corps-ember/pull/653 which got stale and I'll do another PR, but I thought I should do it once the remaining of my PRs are merged.
